tgui/packages/common/color.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file
+ * @copyright 2020 Aleksej Komarov
+ * @license MIT
+ */
+
+const EPSILON = 0.0001;
+
+export class Color {
+  r: number;
+  g: number;
+  b: number;
+  a: number;
+  /**
+   * Creates a color from the CSS hex color notation.
+   */
+  static fromHex: (hex: string) => Color;
+  /**
+   * Linear interpolation of two colors.
+   */
+  static lerp: (c1: Color, c2: Color, n: number) => Color;
+  /**
+   * Loops up the color in the provided list of colors
+   * with linear interpolation.
+   */
+  static lookup: (value: number, colors: Color[]) => Color;
+
+  constructor(r = 0, g = 0, b = 0, a = 1) {
+    this.r = r;
+    this.g = g;
+    this.b = b;
+    this.a = a;
+  }
+
+  toString(): string {
+    // Alpha component needs to permit fractional values, so cannot use |
+    let alpha = this.a;
+    if (typeof alpha === 'string') {
+      alpha = parseFloat(this.a as any);
+    }
+    if (isNaN(alpha)) {
+      alpha = 1;
+    }
+    return `rgba(${this.r | 0}, ${this.g | 0}, ${this.b | 0}, ${alpha})`;
+  }
+
+  /**  Darkens a color by a given percent. Returns a color, which can have toString called to get it's rgba() css value. */
+  darken(percent: number): Color {
+    percent /= 100;
+    return new Color(
+      this.r - this.r * percent,
+      this.g - this.g * percent,
+      this.b - this.b * percent,
+      this.a,
+    );
+  }
+
+  /** Brightens a color by a given percent. Returns a color, which can have toString called to get it's rgba() css value. */
+  lighten(percent: number): Color {
+    // No point in rewriting code we already have.
+    return this.darken(-percent);
+  }
+
+  fromHex = (hex: string): Color => {
```

Comment:
Why not `static fromHex(...) {`?

Also, `new Color().fromHex()` is not how to use a static function, seems wrong.

Reply:
I knew there had to be something wrong there, tyty